File: csharp/decorator.cs
abstract class Coffee {
    public abstract int Cost();
}

class BasicCoffee : Coffee {
    public override int Cost() => 5;
}

abstract class CoffeeDecorator : Coffee {
    protected Coffee coffee;

    public CoffeeDecorator(Coffee coffee) {
        this.coffee = coffee;
    }
}

class MilkDecorator : CoffeeDecorator {
    public MilkDecorator(Coffee coffee) : base(coffee) {}

    public override int Cost() => coffee.Cost() + 2;
}

class SugarDecorator : CoffeeDecorator {
    public SugarDecorator(Coffee coffee) : base(coffee) {}

    public override int Cost() => coffee.Cost() + 1;
}

File: kotlin/decorator.kt
interface Coffee {
    fun cost(): Int
}

class BasicCoffee : Coffee {
    override fun cost() = 5
}

abstract class CoffeeDecorator(private val coffee: Coffee) : Coffee {
    protected fun baseCost() = coffee.cost()
}

class MilkDecorator(coffee: Coffee) : CoffeeDecorator(coffee) {
    override fun cost() = baseCost() + 2
}

class SugarDecorator(coffee: Coffee) : CoffeeDecorator(coffee) {
    override fun cost() = baseCost() + 1
}
